### PR TITLE
Run node docker

### PR DESCRIPTION
**Pull request by JudahSchwartz**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: CIAC-3820
depends on: https://github.com/demisto/dockerfiles/pull/8569

## Description
When node or the necessary modules arent installed, bring up the server in a docker container

## Screenshots
Paste here any images that will help the reviewer

## Must have
- [x] Tests
- [ ] Documentation
- [ ] [VS Code Extension](https://docs.gitlab.com/ee/ci/yaml/#tags)
  - [ ] Functionality implemented in the extension - <u>\<link-to-pr-here\></u>
  - [ ] N/A - Functionality cannot be implemented in the extension because <u>\<add-reason-here\></u>

### CODE REVIEW

File: demisto_sdk/commands/common/MDXServer.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+_MDX_SERVER_PROCESS: Optional[subprocess.Popen] = None
+
+
+def server_script_path():
+    return Path(__file__).parent.parent / 'common' / _SERVER_SCRIPT_NAME
+
+
+class DockerMDXServer:
+    def __enter__(self):
+        image_name = 'devdemisto/demisto-sdk-dependencies:1.0.0.33871'
```
Comment (dorschw) [inline demisto_sdk/commands/common/MDXServer.py:29]: 1. Hard-coded? 
2. Do we ever update it?

File: demisto_sdk/commands/common/MDXServer.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+            files_to_push=[(server_script_path(), location_in_docker)],
+            auto_remove=True,
+            ports={'6161/tcp': ('localhost', 6161)}
+
+        )
+        self._container = container
+        container.start()
+        if 'MDX server is listening on port' not in (str(next(container.logs(stream=True)).decode('utf-8'))):
+            self._container.stop()
+            logging.info('Docker server was not started correctly')
```
Comment (dorschw) [inline demisto_sdk/commands/common/MDXServer.py:49]:
```suggestion
            logging.error('Docker for MDX server was not started correctly')
```

File: demisto_sdk/commands/common/MDXServer.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            logging.info(f'docker logs: {container.logs().decode("utf-8")}')
```
Comment (dorschw) [inline demisto_sdk/commands/common/MDXServer.py:50]:
```suggestion
            logging.error(f'docker logs:\n{container.logs().decode("utf-8")}')
```

File: demisto_sdk/commands/common/errors.py [hunk 4]
```diff
@@ -839,8 +839,8 @@ def readme_missing_output_context(command, context_paths):
 
     @staticmethod
     @error_code_decorator
-    def error_starting_mdx_server(line):
-        return f'Failed starting mdx server. stdout: {line}.\n' \
+    def error_starting_mdx_server():
+        return f'Failed starting mdx server.\n' \
                f'Try running the following command: `npm install`'
```
Comment (dorschw) [inline demisto_sdk/commands/common/errors.py:844]: is it the same error when failing to run locally, and in a docker?

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 5]
```diff
@@ -54,6 +59,51 @@
...
+        2. node interperter exists.
+    Returns:
+        bool: True If all req ok else False
+    """
+    missing_module = []
+    valid = True
+    # Check node exist
+    stdout, stderr, exit_code = run_command_os('node -v', cwd=content_path)
+    if exit_code:
+        print_warning(f'There is no node installed on the machine, Test Skipped, error - {stderr}, {stdout}')
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/readme.py:88]:
```suggestion
        print_warning("Node was not found on the machine - skipping test. error:{stderr}, {stdout}')
```

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 6]
```diff
@@ -54,6 +59,51 @@
...
+        if exit_code:  # all are missinig
+            missing_module.extend(REQUIRED_MDX_PACKS)
+        else:
+            deps = json.loads(stdout).get('dependencies', {})
+            for pack in REQUIRED_MDX_PACKS:
+                if pack not in deps:
+                    missing_module.append(pack)
+    if missing_module:
+        valid = False
+        print_warning(f"The npm modules: {missing_module} are not installed. Use "
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/readme.py:103]: may be more than one, i'd add 
```python
missing_modules_str = ', '.join(sortred(missing_module))
```

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 7]
```diff
@@ -657,7 +676,8 @@ def verify_copyright_section_in_readme_content(self):
             True if related words does not exist in the README content, and False if it does.
         """
         is_valid = True
-        invalid_lines = self.check_readme_content_contain_text(text_list=['BSD', 'MIT', 'Copyright', 'proprietary'], to_split=True)
+        invalid_lines = self.check_readme_content_contain_text(text_list=['BSD', 'MIT', 'Copyright', 'proprietary'],
+                                                               to_split=True)
```
Comment (dorschw) [inline demisto_sdk/commands/common/hook_validations/readme.py:680]: please undo if unintentional

File: demisto_sdk/commands/common/tests/readme_test.py [hunk 8]
```diff
@@ -90,7 +138,8 @@ def test_relative_url_not_valid():
     captured_output = io.StringIO()
     sys.stdout = captured_output  # redirect stdout.
     absolute_urls = ["https://www.good.co.il", "https://example.com", "https://github.com/demisto/content/blob/123",
-                     "github.com/demisto/content/blob/123/Packs/FeedOffice365/doc_files/test.png", "https://hreftesting.com"]
+                     "github.com/demisto/content/blob/123/Packs/FeedOffice365/doc_files/test.png",
+                     "https://hreftesting.com"]
```
Comment (dorschw) [inline demisto_sdk/commands/common/tests/readme_test.py:142]: please undo if unintentional

**Review by dorschw — COMMENTED** [review]
cool!

**Comment by JudahSchwartz** [comment]
reminder to myself to run nightly before merge

File: demisto_sdk/commands/common/MDXServer.py (hunk 1 above)
Comment (JudahSchwartz) [reply demisto_sdk/commands/common/MDXServer.py]: moved to constants. We will probably update with code changes

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 9]
```diff
@@ -657,7 +671,8 @@ def verify_copyright_section_in_readme_content(self):
...
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/common/hook_validations/readme.py:675]:
```suggestion
        invalid_lines = self.check_readme_content_contain_text(text_list=['BSD', 'MIT', 'Copyright', 'proprietary'], to_split=True)
```

File: demisto_sdk/commands/common/tests/readme_test.py [hunk 10]
```diff
@@ -90,7 +134,8 @@ def test_relative_url_not_valid():
...
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/common/tests/readme_test.py:138]:
```suggestion
                     "github.com/demisto/content/blob/123/Packs/FeedOffice365/doc_files/test.png", "https://hreftesting.com"]
```

File: demisto_sdk/tests/workflow_test.py [hunk 11]
```diff
@@ -318,7 +318,7 @@ def rename_incident_field(content_repo: ContentGitRepo, monkeypatch: MonkeyPatch
     modify_entity,
     rename_incident_field
 ])
-def test_workflow_by_sequence(function: Callable, monkeypatch: MonkeyPatch):
+def test_workflow_by_sequence(function: Callable, monkeypatch: MonkeyPatch, mocker):
```
Comment (JudahSchwartz) [inline demisto_sdk/tests/workflow_test.py:321]:
```suggestion
def test_workflow_by_sequence(function: Callable, monkeypatch: MonkeyPatch):
```

File: demisto_sdk/commands/validate/validate_manager.py [hunk 12]
```diff
@@ -381,7 +381,7 @@ def run_validation_on_all_packs(self):
 
         ReadMeValidator.add_node_env_vars()
         if self.is_possible_validate_readme:
-            with ReadMeValidator.start_mdx_server(handle_error=self.handle_error):
+            with ReadMeValidator.start_mdx_server():
```
Comment (JudahSchwartz) [inline demisto_sdk/commands/validate/validate_manager.py:384]:
```suggestion
            with ReadMeValidator.start_mdx_server(handle_error=self.handle_error):
```

File: demisto_sdk/commands/common/hook_validations/readme.py (hunk 6 above)
Comment (JudahSchwartz) [reply demisto_sdk/commands/common/hook_validations/readme.py]: prints as list, that should be built in (also I didnt write this func, only moved it)

**Comment by JudahSchwartz** [comment]
Nightly Run:
https://code.pan.run/xsoar/content/-/pipelines/3742704

File: demisto_sdk/commands/common/hook_validations/readme.py [hunk 13]
```diff
@@ -331,6 +299,49 @@ def check_readme_relative_url_paths(self, is_pack_readme: bool = False) -> list:
 
         return error_list
 
+    @staticmethod
+    @lru_cache(None)
+    def is_docker_available():
```
Comment (ShahafBenYakir) [inline demisto_sdk/commands/common/hook_validations/readme.py:304]: Let's add docstring

File: demisto_sdk/commands/common/MDXServer.py [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+_MDX_SERVER_PROCESS: Optional[subprocess.Popen] = None
+_RUNNING_CONTAINER_IMAGE: Optional[docker.models.containers.Container] = None
+
+
+def server_script_path():
+    return Path(__file__).parent.parent / 'common' / _SERVER_SCRIPT_NAME
+
+
+@contextmanager
+def start_docker_MDX_server(handle_error: Optional[Callable] = None, file_path: Optional[str] = None):
```
Comment (ShahafBenYakir) [inline demisto_sdk/commands/common/MDXServer.py:30]: Let's add docstring

File: demisto_sdk/commands/common/MDXServer.py [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+
+
+def stop_docker_container(container):
+    if container:
+        logging.info('Stopping mdx docker server')
+        container.stop()  # type: ignore
+
+
+@contextmanager
+def start_local_MDX_server(handle_error: Optional[Callable] = None, file_path: Optional[str] = None):
```
Comment (ShahafBenYakir) [inline demisto_sdk/commands/common/MDXServer.py:72]: Let's add docstring

File: demisto_sdk/tests/workflow_test.py [hunk 16]
```diff
@@ -202,6 +202,7 @@ def function_setup():
         content_git_repo = ContentGitRepo()
     # Function setup
     content_git_repo.git_cleanup()
+    content_git_repo.run_command("npm install")
```
Comment (ShahafBenYakir) [inline demisto_sdk/tests/workflow_test.py:205]: Doesn't this mean that for every test we re-install npm packages?

File: tests_end_to_end/test_dockerserver.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
+import requests
+from requests.sessions import HTTPAdapter
+from urllib3 import Retry
+
+from demisto_sdk.commands.common.MDXServer import (DEMISTO_DEPS_DOCKER_NAME,
+                                                   start_docker_MDX_server)
+from demisto_sdk.commands.lint.docker_helper import init_global_docker_client
+
+
+def assert_successful_mdx_call():
```
Comment (ShahafBenYakir) [inline tests_end_to_end/test_dockerserver.py:10]: Let's add docstring to all test functions (given when then). It is important to easily understand what we are testing in each test.

File: tests_end_to_end/test_dockerserver.py [hunk 18]
```diff
@@ -0,0 +1,66 @@
...
+    assert response.status_code == 500
+
+
+def container_is_up():
+    return any(container.name == DEMISTO_DEPS_DOCKER_NAME
+               for container in init_global_docker_client().containers.list())
+
+
+def test_is_file_not_valid():
+    with start_docker_MDX_server() as boolean:
```
Comment (ShahafBenYakir) [inline tests_end_to_end/test_dockerserver.py:44]: Are we actually creating a server while testing? Shouldn't this be mocked?

File: demisto_sdk/commands/common/tests/readme_test.py [hunk 19]
```diff
@@ -48,7 +53,34 @@ def test_is_file_valid(mocker, current, answer):
               status_code=200, text="Test3")
         mocker.patch.dict(os.environ, {'DEMISTO_README_VALIDATION': 'yes', 'DEMISTO_MDX_CMD_VERIFY': 'yes'})
         assert readme_validator.is_valid_file() is answer
-        assert not ReadMeValidator._MDX_SERVER_PROCESS
+        assert not demisto_sdk.commands.common.MDXServer._MDX_SERVER_PROCESS
+
+
+def test_local_server_up_and_down():
```
Comment (ShahafBenYakir) [inline demisto_sdk/commands/common/tests/readme_test.py:59]: Let's add docstring to your new test functions

**Review by ShahafBenYakir — COMMENTED** [review]
Great work! making sure we have tests for the negative cases where either node is not installed, Docker is not installed, or Docker is installed but node is not installed within it.

File: tests_end_to_end/test_dockerserver.py (hunk 18 above)
Comment (JudahSchwartz) [reply tests_end_to_end/test_dockerserver.py:44]: this is under end_to_end, running once during nightly, so I dont think there would be an issue

File: .circleci/config.yml [hunk 20]
```diff
@@ -247,7 +248,9 @@ jobs:
         - attach_workspace:
             at: ~/project
         - <<: *install_build_dependencies
-        - <<: *install_node_ci
+        - node/install-packages:
+            cache-version: << pipeline.parameters.cache-version >>
+            app-dir: content
```
[the remark this answers is not in the document]
Comment (JudahSchwartz) [reply .circleci/config.yml:253]: worked this through with @ilaner , (correct me if Im wrong)
The sdk expects that the npm dependencies are defined in the **content** repo, not demisto-sdk repo,
so this was not properly working before

File: demisto_sdk/tests/workflow_test.py (hunk 16 above)
Comment (JudahSchwartz) [reply demisto_sdk/tests/workflow_test.py:205]: yes I think so. But we're also reinstalling the repo so I think we need to

Comment (JudahSchwartz) [reply demisto_sdk/tests/workflow_test.py:205]: if the repo was taken from cache and these arent installed it will finish quickly

**Comment by JudahSchwartz** [comment]
@ShahafBenYakir node is installed in the docker image out of the box so that last case shouldnt be a concern. The regular unit-tests are cases where docker is not available.
I'll add a case for nothing installed

**Review by ShahafBenYakir — APPROVED** [review]

**Pull request merged**